### ISSUE TITLE
RangeError: Maximum call stack size exceeded with AWS CDK and Poetry

### ISSUE THREAD

**Issue by gshpychka** [microsoft/pyright#3433]
**Describe the bug**
Pyright seems to not be able to handle aws-cdk-lib. It crashes the analysis with a simple repro. Removing the `pyproject.toml` fixes the issue.

**To Reproduce**
Create a virtualenv with poetry (`poetry init`).

My `pyproject.toml`:
```
[tool.poetry]
name = "aws-cdk-pyright"
version = "0.1.0"
description = ""
authors = ["Glib Shpychka <23005347+gshpychka@users.noreply.github.com>"]

[tool.poetry.dependencies]
python = "^3.10"
aws-cdk-lib = "^2.23.0"


[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```
`test.py`:
```python
import aws_cdk as cdk

app = cdk.App()
```
`pyright test.py`
**Expected behavior**
No errors

**Actual behavior**
```
No configuration file found.
pyproject.toml file found at /home/gshpychka/projects/alpha/testing/aws-cdk-pyright.
Loading pyproject.toml file at /home/gshpychka/projects/alpha/testing/aws-cdk-pyright/pyproject.toml
Pyproject file "/home/gshpychka/projects/alpha/testing/aws-cdk-pyright/pyproject.toml" is missing "[tool.pyright]" section.
stubPath /home/gshpychka/projects/alpha/testing/aws-cdk-pyright/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
An internal error occurred while type checking file "/home/gshpychka/projects/alpha/testing/aws-cdk-pyright/app.py": RangeError: Maximum call stack size exceeded
    at Object.resolveAliasDeclaration (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/aliasDeclarationUtils.ts:26:17)
    at resolveAliasDeclaration (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18586:33)
    at getInferredTypeOfDeclaration (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18327:30)
    at forEach (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18721:36)
    at Array.forEach (<anonymous>)
    at getEffectiveTypeOfSymbolForUsage (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18680:15)
    at getTypeOfName (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3811:39)
    at getTypeOfExpression (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:912:30)
    at getTypeOfMemberAccess (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4343:32)
    at Object.Ae [as getTypeOfExpression] (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:917:30)
Error performing analysis: RangeError: Maximum call stack size exceeded
    at Object.resolveAliasDeclaration (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/aliasDeclarationUtils.ts:26:17)
    at resolveAliasDeclaration (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18586:33)
    at getInferredTypeOfDeclaration (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18327:30)
    at forEach (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18721:36)
    at Array.forEach (<anonymous>)
    at getEffectiveTypeOfSymbolForUsage (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18680:15)
    at getTypeOfName (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3811:39)
    at getTypeOfExpression (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:912:30)
    at getTypeOfMemberAccess (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:4343:32)
    at Object.Ae [as getTypeOfExpression] (/usr/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:917:30)
```
**Additional context**

`pyright --version`: `1.1.244`

**Comment by erictraut** [comment]
Thanks for the repro! I really appreciate it.

I was able to find and fix the underlying cause. The issue is that this library declares many hundreds of methods in a single file and uses the same decorator (`jsii.member`) on all of them. The type evaluation algorithm in pyright was recursively evaluating each of these calls, eventually running out of stack space. So this wasn't a case of infinite recursion, but it was ver deep recursion.

I've made several changes to the algorithm that allows us to avoid the deep recursion in this case. This will be fixed in the next release.

**Comment by gshpychka** [comment]
> Thanks for the repro! I really appreciate it.
> 
> 
> 
> I was able to find and fix the underlying cause. The issue is that this library declares many hundreds of methods in a single file and uses the same decorator (`jsii.member`) on all of them. The type evaluation algorithm in pyright was recursively evaluating each of these calls, eventually running out of stack space. So this wasn't a case of infinite recursion, but it was ver deep recursion.
> 
> 
> 
> I've made several changes to the algorithm that allows us to avoid the deep recursion in this case. This will be fixed in the next release.

Thank you! Any reason why removing pyproject.toml fixes the issue?

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.245, which I just published. It will also be included in the next release of pylance.

**Comment by jbaiad** [comment]
@erictraut thank you for the quick fix!
